packages/drivers/routerlicious-driver/src/documentDeltaConnection.ts

```diff
@@ -32,10 +32,17 @@ export class R11sDocumentDeltaConnection extends DocumentDeltaConnection
                     documentId: id,
                     tenantId,
                 },
-                reconnection: false,
-                transports: ["websocket"],
+                // Allow 1 reconnection attempt so that polling can be tried
+                reconnection: true,
+                reconnectionAttempts: 1,
+                // Enable long-polling as a downgrade option
+                transports: ["websocket", "polling"],
                 timeout: timeoutMs,
             });
+        socket.on("connect_error", () => {
+            // Fallback to polling upgrade mechanism on connection failure
+            socket.io.opts.transports = ["polling", "websocket"];
+        });
```

Comment:
Is there a way to know that "connect_error" happened due to websocket not being available and not for any other transport-level failure?

@GaryWilber FYI

Reply:
There is an `err` param passed to this method. It should probably be logged somewhere to help with debugging

Reply:
@tanviraumi just to clarify about websocket availability, I think there are 2 scenarios we are talking about:
1. the client's browser does not support WebSockets (only applies to IE 9-)
2. there is a proxy or firewall between the client that does not allow WebSocket connections

For scenario #1, the initial configuration (`[websocket, polling]`) automatically covers it and uses polling instead.

For scenario #2, I agree that it might be ideal to have more specificity in the error handler. According to the [socket.io connect_error docs](https://socket.io/docs/v4/client-socket-instance/#connect_error):
> This event is fired when:
> - the low-level connection cannot be established
> - the connection is denied by the server in a [middleware function](https://socket.io/docs/v4/middlewares/)
> 
> In the first case, the Socket will automatically try to reconnect, after a given delay.
>
> In the latter case, you need to manually reconnect. You might need to update the credentials:

The R11s implementation of the WebSocket server does not utilize socket.io middleware functions (i.e. `io.use(...)`, so we shouldn't see any errors related to that. Therefore, all connect_errors for our use-case will be in the "low-level connection cannot be established" category. My test repro of this is limited to a hacky way of simulating the transport-layer failure by closing a WebSocket connection before a handshake is able to be made, but the `err` param @GaryWilber mentioned has some useful information (not sure if we can guarantee this all will stay the same across socket.io versions and browsers, other than `TransportError` type which is [hardcoded in engine.io-client for all low-level errors](https://github.com/socketio/engine.io-client/blob/3.5.x/lib/transport.js#L69)):
1. WebSocket connection fails (via my hack): `{ type: "TransportError", message: "websocket error", description: [Event] }`
    - Event here is an object that I can't discern any helpful information from in the console
2. Polling connection fails (disallowed polling on the server): `{ type: "TransportError", message: "xhr poll error", description: 400 }`
    - The description corresponds to the HTTP 400 error containing `{ code: 0, description: "Transport unknown" }`  caused by me disallowing polling on the server

Reply:
Thanks to @vladsud's suggestion, I was able to more realistically replicate the scenario where a proxy or firewall prevents WebSocket connection upgrade by adding NGINX in front of Alfred locally, then dropping the Upgrade and Connection HTTP headers in order to disable WebSocket compatibility. This allowed me to easily confirm that the `TransportError` with message `"websocket error"` is consistent across Safari, Firefox, and Chromium, as well as the description being an Event object.

I've pushed a commit to take this information into account, please take a look

Reply:
Happy to be useful sometimes :)